src/python/pants/engine/internals/defaults.py

```diff
@@ -0,0 +1,190 @@
+                    }
+                )
+                for target_alias, fields in self.defaults.items()
+            }
+        )
+
+    def get(self, target_alias: str) -> Mapping[str, Any]:
+        # Used by `pants.engine.internals.parser.Parser._generate_symbols.Registrar.__call__`
+        return self.defaults.get(target_alias, {})
+
+    def set_defaults(
+        self,
+        *args: SetDefaultsT,
+        all: SetDefaultsValueT | None = None,
+        extend: bool = False,
+        **kwargs,
+    ) -> None:
+        defaults: dict[str, dict[str, Any]] = (
+            {} if not extend else {k: dict(v) for k, v in self.defaults.items()}
+        )
+
+        if all is not None:
+            self._process_defaults(
+                defaults,
+                {tuple(self.registered_target_types.aliases): all},
+                ignore_unknown_fields=True,
+            )
+
+        for arg in args:
+            self._process_defaults(defaults, arg)
+
+        # Update with new defaults, dropping targets without any default values.
+        for tgt, default in defaults.items():
+            if not default:
+                self.defaults.pop(tgt, None)
+            else:
+                self.defaults[tgt] = default
+
+    def _process_defaults(
+        self,
+        defaults: dict[str, dict[str, Any]],
+        targets_defaults: SetDefaultsT,
+        ignore_unknown_fields: bool = False,
+    ):
+        if not isinstance(targets_defaults, dict):
+            raise ValueError(
+                f"Expected dictionary mapping targets to default field values for {self.address} "
+                f"but got: {type(targets_defaults).__name__}."
+            )
+
+        types = self.registered_target_types.aliases_to_types
+        for target, default in targets_defaults.items():
+            if not isinstance(default, dict):
+                raise ValueError(
+                    f"Invalid default field values in {self.address} for target type {target}, "
+                    f"must be an `dict` but was {default!r} with type `{type(default).__name__}`."
+                )
+
+            targets: Iterable[str]
+            targets = target if isinstance(target, tuple) else (target,)
+            for target_alias in map(str, targets):
+                if target_alias in types:
+                    target_type = types[target_alias]
+                else:
+                    raise ValueError(f"Unrecognized target type {target_alias} in {self.address}.")
+
+                # Copy default dict if we may mutate it.
+                raw_values = dict(default) if ignore_unknown_fields else default
+
+                # Validate that field exists on target
+                target_fields = target_type.class_field_types(self.union_membership)
+                valid_field_aliases = set()
+
+                # TODO: this valid aliases calculation is done every time a target is instantiated
```

Comment:
Can you add a comment near the other location where it is done referring to this case, and the possibility of deduping it?

Reply:
Oh, right. That's been fixed already, should adapt here: https://github.com/pantsbuild/pants/pull/15861